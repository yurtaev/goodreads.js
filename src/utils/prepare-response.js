// @flow

import { propEq, cond, T, propSatisfies } from 'ramda'

import type { ResponseAPI } from './../../types/api'

export const isOK = propEq('ok', true)
export const isNotOK = propEq('ok', false)

export const is20x = propSatisfies(status => status >= 200 && status <= 299, 'status')
export const is40x = propSatisfies(status => status >= 400 && status <= 499, 'status')
export const is50x = propSatisfies(status => status >= 500 && status <= 599, 'status')

const on20x = async (response: Response) => {
  const result = await response.text()
  return {
    ok: true,
    status: response.status,
    data: result,
  }
}

const on50x = async (response: Response) => {
  const text = await response.text()
  return Promise.reject({
    ok: false,
    statusText: response.statusText,
    status: response.status,
    error: text,
  })
}

const on40x = async (response: Response) => {
  const result = await response.text()
  return Promise.reject({
    ok: false,
    status: response.status,
    error: result,
  })
}

const onUnknow = (response: Response) =>
  Promise.reject({
    ok: false,
    status: response.status,
    statusText: response.statusText,
    error: 'Unknown response status',
  })

const onOK = async (response: Response) => cond([[is20x, on20x]])(response)
const onNotOK = async (response: Response) =>
  cond([[is40x, on40x], [is50x, on50x], [T, onUnknow]])(response)

export const prepareResponse = async (response: Response): Promise<ResponseAPI<any>> =>
  cond([[isOK, onOK], [isNotOK, onNotOK]])(response)
