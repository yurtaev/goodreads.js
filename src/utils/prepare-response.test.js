// @flow

import { failure } from 'awaiting'

import { prepareResponse } from './prepare-response'

const dataOK = "xml data"
const dataError = "response error"

describe('prepareResponse', () => {
  test('200', async () => {
    const response = new Response(dataOK, {
      ok: true,
      status: 200,
    })

    const resp = await prepareResponse(response)

    expect(resp).toMatchSnapshot('200')
  })

  test('400', async () => {
    const response = new Response(dataError, {
      ok: false,
      status: 400,
    })

    const err = await failure(prepareResponse(response))
    expect(err).toMatchSnapshot('400')
  })

  test('401', async () => {
    const response = new Response(dataError, {
      ok: false,
      status: 401,
    })

    const err = await failure(prepareResponse(response))
    expect(err).toMatchSnapshot('401')
  })

  test('500', async () => {
    const response = new Response(dataError, {
      ok: false,
      status: 500,
      statusText: 'internal server error',
    })

    const err = await failure(prepareResponse(response))
    expect(err).toMatchSnapshot('500')
  })

  test('Unknown response status', async () => {
    const response = new Response(dataError, {
      ok: false,
      status: 500,
      statusText: 'Unknown response status',
    })

    const err = await failure(prepareResponse(response))
    expect(err).toMatchSnapshot('500')
  })
})
