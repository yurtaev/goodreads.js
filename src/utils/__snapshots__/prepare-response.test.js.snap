// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`200 1`] = `
Object {
  "data": "xml data",
  "ok": true,
  "status": 200,
}
`;

exports[`400 1`] = `
Object {
  "error": "response error",
  "ok": false,
  "status": 400,
}
`;

exports[`401 1`] = `
Object {
  "error": "response error",
  "ok": false,
  "status": 401,
}
`;

exports[`500 1`] = `
Object {
  "error": "response error",
  "ok": false,
  "status": 500,
  "statusText": "internal server error",
}
`;

exports[`500 2`] = `
Object {
  "error": "response error",
  "ok": false,
  "status": 500,
  "statusText": "Unknown response status",
}
`;
